File: Solutions/941-Valid Mountain Array/941-Valid Mountain Array.rb
#Third fastest
def valid_mountain_array(a)
    return false if a.size < 3
    direction = 1
    for i in 0..a.size-1
        if i > 0
            return false if a[i] - a[i-1] == 0
            direction = 0 if a[i] - a[i-1] < 0
        else
            return false if a[i] - a[i+1] > 0
        end
        return false if a[i] - a[i-1] > 0 if direction == 0
    end
    return direction == 0
end

#Second fastest
def valid_mountain_array(a)
    return false if a.length < 3
    l = left(a)
    r = right(a)
    !l || !r ? false : l == r
end

def left(a)
    i = 0
    while i < a.length - 1 do
        return false if a[i] == a[i+1]
        break if a[i] > a[i+1]
        i+=1
    end
    return i == a.length - 1 ? false : i
end

def right(a)
    i = a.length - 1
    while i > 0 do
        return false if a[i] == a[i-1]
        break if a[i-1] < a[i]
        i-=1
    end
    return i == 0 ? false : i
end

#Fastest
def valid_mountain_array(arr)
  # return false if size < 3 OR array starts in decreasingorder
  return false if arr.size < 3 || arr[0] > arr[1]

  dec_found = false
  i = 1
  
  # traverse the increasing part
  while i < arr.size
    return false if arr[i] == arr[i-1]
    if arr[i] < arr[i-1]
      dec_found = true
      break
    end
    i += 1
  end

  
  # finished iteration but dec_found = false
  # which means only increasing values present
  # so return false
  return false unless dec_found

# traverse the decreasing part of the array
  while i < arr.size
    return false if arr[i] >= arr[i-1]
    i += 1
  end

  true
end

File: Solutions/941-Valid Mountain Array/941-Valid Mountain Array.cs
public class Solution {
    public bool ValidMountainArray(int[] arr) {
        if(arr.Length < 3)
            return false;
        int maxIdx = 0;
        for(int i = 1; i < arr.Length; i++)
            if(arr[i] > arr[maxIdx])
                maxIdx = i;
        
        int a = maxIdx - 1;
        int b = maxIdx + 1;
        
        if(a < 0 || b > arr.Length-1)
            return false;
        
        while(a >= 0){
            if(arr[a] >= arr[a+1])
                return false;
            a--;
        }
        
        while(b < arr.Length){
            if(arr[b-1] <= arr[b])
                return false;
            b++;
        }
        
        return true;
    }
}

//Third fastest
public class Solution {
    public bool ValidMountainArray(int[] arr) {
        if(arr.Length < 3) return false;
        bool isClimbed = false, isClimbing = true;
        for(int i = 1; i< arr.Length; ++i){
            if(arr[i - 1] < arr[i]){
                if(!isClimbed) isClimbed = true;
                if(!isClimbing) return false;
            } else if(arr[i - 1] > arr[i]){
                if(!isClimbed) return false;
                if(isClimbing) {
                    isClimbing = false;
                }
            } else{
                return false;
            }
        }
        return isClimbing == false;
    }
}

//Second fastest
public class Solution {
    public bool ValidMountainArray(int[] arr) {
        if(arr.Length<3){
            return false;
        }
        bool isMountain = true;
        int i = 0;
        bool reachedPeak = false; 
        while(isMountain && ++i<arr.Length-1){
            if(!reachedPeak){
                if(arr[i]>arr[i-1] && arr[i]>arr[i+1]){
                    reachedPeak = true; 
                }
                else if(!(arr[i]>arr[i-1] && arr[i]<arr[i+1])) {
                    isMountain = false;
                }
            }else{
                if(!(arr[i]<arr[i-1] && arr[i]>arr[i+1])){
                    isMountain = false;
                }
            }

           
        }
        return isMountain && reachedPeak;
    }
                   
}

//Fastest
public class Solution 
{
    public bool ValidMountainArray(int[] arr)
    {
        if(arr == null || arr.Length < 3) return false;
        
        int left = 0;
        int right = arr.Length-1;
        
        while(left < right)
        {
            if(arr[left] < arr[left+1])
            {
                left++;
            }
            else if(arr[right-1] > arr[right] )
            {
                right--;
            }
            else
            {
                break;
            }
        }
        
        if(left == right && left != 0 && right != arr.Length-1)
        {
            return true;
        }
        
        return false;
    }
}